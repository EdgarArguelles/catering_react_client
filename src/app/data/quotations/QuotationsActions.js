/* eslint-disable max-lines */
/**
 * Following the Flux Standard Action: https://github.com/acdlite/flux-standard-action
 */
import keyMirror from 'keymirror';
import Api from 'app/common/Api';
import Utils from 'app/common/Utils';

export const ACTION_TYPES = keyMirror({
  QUOTATIONS_CLEAN_ERROR: null,
  QUOTATIONS_CLEAN_QUOTATIONS: null,
  QUOTATION_OVERWRITE_LOCAL: null,
  FETCH_QUOTATIONS_REQUEST: null,
  FETCH_QUOTATIONS_SUCCESS: null,
  FETCH_QUOTATIONS_ERROR: null,
  FETCH_QUOTATION_REQUEST: null,
  FETCH_QUOTATION_SUCCESS: null,
  FETCH_QUOTATION_ERROR: null,
  CREATE_QUOTATIONS_REQUEST: null,
  CREATE_QUOTATIONS_SUCCESS: null,
  CREATE_QUOTATIONS_ERROR: null,
  EDIT_QUOTATIONS_REQUEST: null,
  EDIT_QUOTATIONS_SUCCESS: null,
  EDIT_QUOTATIONS_ERROR: null,
  DELETE_QUOTATIONS_REQUEST: null,
  DELETE_QUOTATIONS_SUCCESS: null,
  DELETE_QUOTATIONS_ERROR: null,
});

/**
 * Clean extra information from quotation (not needed in graphql mutation)
 *
 * @param {Object} quotation data to be cleaned
 * @return {String} data cleaned
 */
const getQuotationFixed = quotation => {
  const quotationFixed = {...quotation, menus: []};
  quotation.menus && quotation.menus.forEach(menu => quotationFixed.menus.push({...menu}));
  quotationFixed.menus.forEach(menu => {
    delete menu.price;
    delete menu.isSelected;
  });

  return Utils.stringifyObjectWithNoQuotesOnKeys(quotationFixed);
};

export default class QuotationsActions {
  static cleanError() {
    return {type: ACTION_TYPES.QUOTATIONS_CLEAN_ERROR};
  }

  static cleanQuotations() {
    return {type: ACTION_TYPES.QUOTATIONS_CLEAN_QUOTATIONS};
  }

  static fetchQuotations(pagination) {
    const FIELDS = 'totalElements totalPages content{id name createdAt price}';
    const body = {query: `{quotationPage(pageDataRequest: ${Utils.stringifyPageDataRequest(pagination)}) {${FIELDS}}}`};

    return async dispatch => {
      try {
        dispatch({type: ACTION_TYPES.FETCH_QUOTATIONS_REQUEST});
        const json = await Api.graphql(dispatch, body);
        const {content, totalElements, totalPages} = json.data.quotationPage;
        dispatch({
          type: ACTION_TYPES.FETCH_QUOTATIONS_SUCCESS,
          payload: {
            data: content,
            metaData: {pagination: {...pagination}, totalElements, totalPages},
          },
        });
        return json;
      } catch (error) {
        dispatch({
          type: ACTION_TYPES.FETCH_QUOTATIONS_ERROR,
          error: true,
          payload: error,
        });
        throw error;
      }
    };
  }

  static fetchQuotation(quotationId, overwriteLocalChanges = true) {
    const FIELDS = 'id name createdAt price menus{id name quantity courses{id position type{id} dishes{id price}}}';
    const body = {query: `{quotation(id: ${quotationId}) {${FIELDS}}}`};

    return async dispatch => {
      try {
        dispatch({type: ACTION_TYPES.FETCH_QUOTATION_REQUEST});
        const json = await Api.graphql(dispatch, body);
        dispatch({
          type: ACTION_TYPES.FETCH_QUOTATION_SUCCESS,
          payload: {
            data: json.data.quotation,
          },
        });
        if (overwriteLocalChanges) {
          dispatch({
            type: ACTION_TYPES.QUOTATION_OVERWRITE_LOCAL,
            payload: {
              data: json.data.quotation,
            },
          });
        }
        return json;
      } catch (error) {
        dispatch({
          type: ACTION_TYPES.FETCH_QUOTATION_ERROR,
          error: true,
          payload: error,
        });
        throw error;
      }
    };
  }

  static createQuotation(quotation) {
    const body = {query: `mutation {createQuotation(quotation: ${getQuotationFixed(quotation)}) {id}}`};

    return async dispatch => {
      try {
        dispatch({type: ACTION_TYPES.CREATE_QUOTATIONS_REQUEST});
        // create a fake delay (ignore it in test cases)
        process.env.NODE_ENV !== 'test' && await new Promise(resolve => setTimeout(resolve, 3000));
        const json = await Api.graphql(dispatch, body);
        dispatch({
          type: ACTION_TYPES.CREATE_QUOTATIONS_SUCCESS,
          payload: {
            data: json.data.createQuotation,
          },
        });
        return json;
      } catch (error) {
        dispatch({
          type: ACTION_TYPES.CREATE_QUOTATIONS_ERROR,
          error: true,
          payload: error,
        });
        throw error;
      }
    };
  }

  static editQuotation(quotation) {
    const body = {query: `mutation {updateQuotation(quotation: ${getQuotationFixed(quotation)}) {id}}`};

    return async dispatch => {
      try {
        dispatch({type: ACTION_TYPES.EDIT_QUOTATIONS_REQUEST});
        // create a fake delay (ignore it in test cases)
        process.env.NODE_ENV !== 'test' && await new Promise(resolve => setTimeout(resolve, 3000));
        const json = await Api.graphql(dispatch, body);
        dispatch({
          type: ACTION_TYPES.EDIT_QUOTATIONS_SUCCESS,
          payload: {
            data: json.data.updateQuotation,
          },
        });
        return json;
      } catch (error) {
        dispatch({
          type: ACTION_TYPES.EDIT_QUOTATIONS_ERROR,
          error: true,
          payload: error,
        });
        throw error;
      }
    };
  }

  static deleteQuotation(quotationId) {
    const body = {query: `mutation {deleteQuotation(id: ${quotationId}) {id}}`};

    return async dispatch => {
      try {
        dispatch({type: ACTION_TYPES.DELETE_QUOTATIONS_REQUEST});
        // create a fake delay (ignore it in test cases)
        process.env.NODE_ENV !== 'test' && await new Promise(resolve => setTimeout(resolve, 3000));
        const json = await Api.graphql(dispatch, body);
        dispatch({
          type: ACTION_TYPES.DELETE_QUOTATIONS_SUCCESS,
          payload: {
            data: json.data.deleteQuotation,
          },
        });
        return json;
      } catch (error) {
        dispatch({
          type: ACTION_TYPES.DELETE_QUOTATIONS_ERROR,
          error: true,
          payload: error,
        });
        throw error;
      }
    };
  }
}