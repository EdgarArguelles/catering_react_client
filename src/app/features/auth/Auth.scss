@use 'app/Variables';

#auth {
  .waiting {
    display: inline-grid;

    .progress {
      margin: -25px auto 0 auto;
      width: 100px;
    }
  }
}