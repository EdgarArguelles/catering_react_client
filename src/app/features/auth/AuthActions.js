/**
 * Following the Flux Standard Action: https://github.com/acdlite/flux-standard-action
 */
import keyMirror from 'keymirror';
import Api from '../../common/Api';

export const ACTION_TYPES = keyMirror({
  SOCKET_CONNECTED: null,
  LOGOUT: null,
  LOGIN_SUCCESS: null,
  PING_USER_SUCCESS: null,
  PING_USER_ERROR: null,
});

/**
 * Remove "accessToken" and "userImage" from sessionStorage
 *
 * @return {void}
 */
const removeToken = () => {
  window.sessionStorage.removeItem('accessToken');
  window.sessionStorage.removeItem('userImage');
};

/**
 * Store "accessToken" and "userImage" in sessionStorage
 *
 * @param {Object} loggedUser user information
 * @return {void}
 */
const saveToken = loggedUser => {
  window.sessionStorage.setItem('accessToken', loggedUser.token);
  loggedUser.image && window.sessionStorage.setItem('userImage', loggedUser.image);
};

export default class AuthActions {
  static connectSocket() {
    return {
      type: ACTION_TYPES.SOCKET_CONNECTED,
    };
  }

  static logout() {
    removeToken();
    return {
      type: ACTION_TYPES.LOGOUT,
    };
  }

  static login(body) {
    saveToken(body.loggedUser);
    return {
      type: ACTION_TYPES.LOGIN_SUCCESS,
      payload: {
        loggedUser: body.loggedUser,
      },
    };
  }

  static fetchPing() {
    const body = {query: '{ping {id fullName image role token permissions}}'};

    return async dispatch => {
      try {
        const json = await Api.graphql(dispatch, body);
        const loggedUser = json && json.data && json.data.ping && json.data.ping.id ? json.data.ping : null;
        if (loggedUser) {
          saveToken(loggedUser);
          dispatch({
            type: ACTION_TYPES.PING_USER_SUCCESS,
            payload: {
              loggedUser,
            },
          });
        } else {
          removeToken();
          dispatch({type: ACTION_TYPES.PING_USER_ERROR});
        }
        return json;
      } catch (error) {
        removeToken();
        dispatch({type: ACTION_TYPES.PING_USER_ERROR});
        throw error;
      }
    };
  }
}