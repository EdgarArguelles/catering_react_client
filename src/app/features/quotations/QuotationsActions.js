/**
 * Following the Flux Standard Action: https://github.com/acdlite/flux-standard-action
 */
import keyMirror from 'keymirror';

export const ACTION_TYPES = keyMirror({
  QUOTATIONS_CHANGE_MENU_TAB: null,
  QUOTATIONS_CHANGE_IS_MENU_DIALOG_OPEN: null,
  QUOTATIONS_DELETE_LOCAL: null,
  QUOTATIONS_END_REMOTE_PROCESS: null,
});

export default class QuotationsActions {
  static changeMenuTab(tab) {
    return {
      type: ACTION_TYPES.QUOTATIONS_CHANGE_MENU_TAB,
      payload: {
        tab,
      },
    };
  }

  static changeMenuDialogOpen(isMenuDialogOpen) {
    return {
      type: ACTION_TYPES.QUOTATIONS_CHANGE_IS_MENU_DIALOG_OPEN,
      payload: {
        isMenuDialogOpen,
      },
    };
  }

  static deleteLocal() {
    return {type: ACTION_TYPES.QUOTATIONS_DELETE_LOCAL};
  }

  static endRemoteProcess() {
    return {type: ACTION_TYPES.QUOTATIONS_END_REMOTE_PROCESS};
  }
}