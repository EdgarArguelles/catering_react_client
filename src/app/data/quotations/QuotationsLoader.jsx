import React, {useEffect} from 'react';
import PropTypes from 'prop-types';
import {useDispatch, useSelector} from 'react-redux';
import {fetchCompleteQuotation} from 'app/features/quotations/quotation/Quotation.service';
import QuotationsActions from './QuotationsActions';

const QuotationsLoader = ({children}) => {
  const dispatch = useDispatch();
  const loggedUser = useSelector(state => state.auth.loggedUser);
  const quotation = useSelector(state => state.quotations.quotation);
  const isFetching = useSelector(state => state.data.fetching.quotations || state.data.fetching.quotationsUpdate);
  const quotations = useSelector(state => state.data.quotations);

  useEffect(() => {
    if (loggedUser) {
      const fetchQuotation = quotationId => dispatch(QuotationsActions.fetchQuotation(quotationId, false));
      fetchCompleteQuotation(quotation, isFetching, quotations, fetchQuotation);
    }
  }, [loggedUser, quotation, isFetching, quotations, dispatch]);

  return <>{children}</>;
};

QuotationsLoader.propTypes = {
  children: PropTypes.object.isRequired,
};

export default React.memo(QuotationsLoader);