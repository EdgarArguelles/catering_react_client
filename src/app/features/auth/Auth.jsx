import './Auth.scss';
import loading from 'assets/img/loading.gif';
import SockJS from 'sockjs-client';
import Stomp from 'stompjs';
import React, {useEffect} from 'react';
import PropTypes from 'prop-types';
import {useDispatch, useSelector} from 'react-redux';
import Facebook from './auth_button/facebook/Facebook.react';
import Google from './auth_button/google/Google.react';
import AuthActions from './AuthActions';

let stompClient;

const useCreateStompClient = (socketConnected, dispatch) => {
  useEffect(() => {
    if (socketConnected) {
      return undefined;
    }

    let timeout = null;
    const createSocket = () => {
      const socket = new SockJS(`${process.env.API_URL}/oauth/websocket`);

      stompClient = Stomp.Stomp.over(socket);
      stompClient.debug = null; // disable console messages in browser
      clearTimeout(timeout);
      stompClient.connect({}, () => dispatch(AuthActions.connectSocket()),
        () => (timeout = setTimeout(() => createSocket(), 5000)));
    };

    createSocket();

    // Clean timeout
    return () => clearTimeout(timeout);
  }, [socketConnected, dispatch]);
};

const Auth = ({onSuccess}) => {
  const dispatch = useDispatch();
  const socketConnected = useSelector(state => state.auth.socketConnected);
  useCreateStompClient(socketConnected, dispatch);

  const subscribe = state => {
    return stompClient.subscribe(`/oauth/response/${state}`, response => {
      dispatch(AuthActions.login(JSON.parse(response.body)));
      onSuccess();
    });
  };

  const getWaiting = () => {
    return (
      <div className="waiting">
        <img src={loading} className="progress" alt="progress"/>
        Conectando...
      </div>
    );
  };

  const getButtons = () => {
    return (
      <>
        <Facebook subscribe={subscribe}/>
        <Google subscribe={subscribe}/>
      </>
    );
  };

  return <div id="auth">{socketConnected ? getButtons() : getWaiting()}</div>;
};

Auth.propTypes = {
  onSuccess: PropTypes.func.isRequired,
};

export default React.memo(Auth);