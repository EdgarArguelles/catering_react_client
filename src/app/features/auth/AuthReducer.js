/**
 * Given the same arguments, it should calculate the next state and return it.
 * No surprises. No side effects. No API calls. No mutations. Just a calculation.
 */
import {ACTION_TYPES as API_ACTION_TYPES} from '../../common/Api';
import {ACTION_TYPES} from './AuthActions';

const socketConnected = (state = false, action) => {
  switch (action.type) {
    case ACTION_TYPES.SOCKET_CONNECTED:
      return true;
    default:
      return state;
  }
};

const loggedUser = (state = null, action) => {
  switch (action.type) {
    case API_ACTION_TYPES.SESSION_EXPIRED:
    case ACTION_TYPES.PING_USER_ERROR:
    case ACTION_TYPES.LOGOUT:
      return null;
    case ACTION_TYPES.LOGIN_SUCCESS:
    case ACTION_TYPES.PING_USER_SUCCESS:
      return action.payload.loggedUser;
    default:
      return state;
  }
};

export default (state = {}, action = {}) => {
  return {
    socketConnected: socketConnected(state.socketConnected, action),
    loggedUser: loggedUser(state.loggedUser, action),
  };
};