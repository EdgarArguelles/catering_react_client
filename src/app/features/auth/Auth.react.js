import './Auth.scss';
import loading from 'assets/img/loading.gif';
import SockJS from 'sockjs-client';
import Stomp from 'stompjs';
import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import Facebook from './auth_button/facebook/Facebook.react';
import Google from './auth_button/google/Google.react';
import AuthActions from './AuthActions';

let stompClient;

class Auth extends React.Component {
  static propTypes = {
    socketConnected: PropTypes.bool.isRequired,
    connectSocket: PropTypes.func.isRequired,
    login: PropTypes.func.isRequired,
    onSuccess: PropTypes.func.isRequired,
  };

  constructor(props) {
    super(props);
    this.timeout = {};
  }

  componentDidMount() {
    const {socketConnected} = this.props;
    !socketConnected && this.createSocket();
  }

  componentWillUnmount() {
    clearTimeout(this.timeout);
    this.timeout = null;
  }

  createSocket = () => {
    const {connectSocket} = this.props;
    const socket = new SockJS(`${process.env.API_URL}/oauth/websocket`);

    stompClient = Stomp.Stomp.over(socket);
    stompClient.debug = null; // disable console messages in browser
    clearTimeout(this.timeout);
    stompClient.connect({}, () => connectSocket(), () => (this.timeout = setTimeout(() => this.createSocket(), 5000)));
  };

  subscribe = state => {
    const {login, onSuccess} = this.props;

    return stompClient.subscribe(`/oauth/response/${state}`, response => {
      login(JSON.parse(response.body));
      onSuccess();
    });
  };

  getWaiting = () => {
    return (
      <div className="waiting">
        <img src={loading} className="progress"/>
        Conectando...
      </div>
    );
  };

  getButtons = () => {
    return (
      <React.Fragment>
        <Facebook subscribe={this.subscribe}/>
        <Google subscribe={this.subscribe}/>
      </React.Fragment>
    );
  };

  render() {
    const {socketConnected} = this.props;

    return <div id="auth">{socketConnected ? this.getButtons() : this.getWaiting()}</div>;
  }
}

const mapStateToProps = state => {
  return {
    socketConnected: state.auth.socketConnected,
  };
};

const mapDispatchToProps = dispatch => {
  return {
    connectSocket: () => {
      dispatch(AuthActions.connectSocket());
    },
    login: body => {
      dispatch(AuthActions.login(body));
    },
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(Auth);