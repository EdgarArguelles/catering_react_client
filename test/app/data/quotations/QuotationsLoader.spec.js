/* eslint-disable max-lines */
import sinon from 'sinon';
import renderer, {act} from 'react-test-renderer';
import React from 'react';
import {Provider} from 'react-redux';
import configureStore from 'redux-mock-store';
import {createStore} from 'redux';
import reducers from 'app/Reducers';
import QuotationsLoader from 'app/data/quotations/QuotationsLoader';
import QuotationsActions, {ACTION_TYPES} from 'app/data/quotations/QuotationsActions';
import {ACTION_TYPES as AUTH_TYPES} from 'app/features/auth/AuthActions';

describe('Data -> Quotations -> Loader', () => {
  const dispatchStub = sinon.stub();
  const fetchQuotationStub = sinon.stub(QuotationsActions, 'fetchQuotation');
  let component, store, wrapper;

  afterEach(() => {
    dispatchStub.reset();
    fetchQuotationStub.reset();
  });

  const mountComponent = (state, useRealStore) => {
    store = configureStore()(state);
    store.dispatch = dispatchStub;
    store = useRealStore ? createStore(reducers, state) : store;
    wrapper = renderer.create(<Provider store={store}>
      <QuotationsLoader><h1>Test</h1></QuotationsLoader>
    </Provider>);
    component = wrapper.root.find(el => el.type.name === 'QuotationsLoader');
  };

  describe('load', () => {
    it('should have all props', () => {
      mountComponent({
        auth: {loggedUser: {id: 'user 1'}},
        quotations: {quotation: {id: 'Q1'}},
        data: {
          fetching: {quotations: true},
          quotations: {Q2: {id: 'Q2'}, Q3: {id: 'Q3'}},
        },
      });

      expect(component.props.children).toStrictEqual(<h1>Test</h1>);
    });
  });

  describe('useEffect -> fetchQuotation', () => {
    beforeEach(() => {
      fetchQuotationStub.withArgs('Q1', false).returns({type: 'TEST'});
      mountComponent({
        auth: {},
        quotations: {quotation: {id: 'Q1'}},
        data: {
          fetching: {quotationsUpdate: true},
        },
      }, true);
    });

    it('should not call fetchQuotation', () => {
      sinon.assert.callCount(fetchQuotationStub, 0);
      sinon.assert.callCount(dispatchStub, 0);
    });

    it('should not call fetchQuotation when dispatch CREATE_QUOTATIONS_SUCCESS', () => {
      act(() => {
        store.dispatch({type: ACTION_TYPES.CREATE_QUOTATIONS_SUCCESS});
      });

      sinon.assert.callCount(fetchQuotationStub, 0);
      sinon.assert.callCount(dispatchStub, 0);
    });

    it('should not call fetchQuotation when loggedUser is present', () => {
      act(() => {
        store.dispatch({
          type: AUTH_TYPES.LOGIN_SUCCESS,
          payload: {
            loggedUser: {id: 'User 2'},
          },
        });
        store.dispatch({
          type: AUTH_TYPES.LOGIN_SUCCESS,
          payload: {
            loggedUser: {id: 'User 3'},
          },
        });
      });

      sinon.assert.callCount(fetchQuotationStub, 0);
      sinon.assert.callCount(dispatchStub, 0);
    });

    it('should call fetchQuotation', () => {
      act(() => {
        store.dispatch({type: ACTION_TYPES.CREATE_QUOTATIONS_SUCCESS});
        store.dispatch({
          type: AUTH_TYPES.LOGIN_SUCCESS,
          payload: {
            loggedUser: {id: 'User 2'},
          },
        });
      });

      sinon.assert.callCount(fetchQuotationStub, 1);
      sinon.assert.calledWithExactly(fetchQuotationStub, 'Q1', false);
      sinon.assert.callCount(dispatchStub, 0);
    });
  });

  describe('snapshot', () => {
    it('should render h1 Test', () => {
      fetchQuotationStub.withArgs('Q1', false).returns({type: 'TEST'});
      mountComponent({
        auth: {},
        quotations: {quotation: {id: 'Q1'}},
        data: {
          fetching: {quotationsUpdate: true},
        },
      });

      global.expect(wrapper.toJSON()).toMatchSnapshot();
    });
  });
});