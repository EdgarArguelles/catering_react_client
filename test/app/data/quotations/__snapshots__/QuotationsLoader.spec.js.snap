// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data -> Quotations -> Loader snapshot should render h1 Test 1`] = `
<h1>
  Test
</h1>
`;
